### PR TITLE
insights: OOB migration from settings to codeinsights-db

### LINKED ISSUE #26533: insights: one time migration for all insights
Currently, insights migrate on a schedule every 10 minutes from settings to the DB. In order to deprecate this process we need to disable these migrations and perform a one time out of band migration. In this OOB migration we should do the following:

1. In global settings
    1. Determine if global has any insights (any kind) defined in the settings
    2. If so, migrate each insight to the database
    3. If so, create a private dashboard with global grant named `Global insights` and attach all insights found in the global settings.
    4. Create any custom dashboards that do not exist, and attach any insights that are associated. These dashboards should have `global` grant
2. For each org
    1. Determine if they have any insights (any kind) defined in their settings
    2. If so, migrate each insight to the database
    3. If so, create a private dashboard with the org grant named `${org name}'s insights` and attach all insights found in the org settings 
    4. Create any custom dashboards that do not exist, and attach any insights that are associated. These dashboards should have the `org` grant
3. For each user
    1. Determine if they have any insights (any kind) defined in their settings
    2. If so, migrate each insight to the database
    3. If so, create a private dashboard with the user grant named `${username}'s insights` and attach all insights found in their settings 
    4. Create any custom dashboards that do not exist, and attach any insights that are associated. These dashboards should have the `user` grant

As long as we go from global -> org -> user we should never find a situation where an insight on a dashboard cannot be referenced.

### PR DESCRIPTION
**Pull request by CristinaBirkel**
Closes #26533, #27306

This is a the feature branch for the OOB migration from settings to the codeinsights-db.

Tasks:

- [x] In band migrations
- [x] Implement Progress function
- [x] Move the store and tables to primary postgres instead of codeinsights-db
- [x] Create special case dashboard for user / org / global
- [x] Better error handling
- [x] Check transactionality of the migration
- [x] Duplicate dashboard check
- [x] Disable 10 minute sync jobs https://github.com/sourcegraph/sourcegraph/pull/27857
- [ ] Test on dogfood and make sure it works with concurrency
- [x] Test plan for edge cases (like non-globally-unique ids) and verify those
- [x] Bug found: it's reusing a BE series for a FE series. Double-check that series re-use logic.
- [x] Log schema errors before skipping them
- [x] BEFORE MERGE: Verify migration numbers are still correct (for both migration files and OOB migration)
- [x] BEFORE MERGE: Verify minor version number in OOB migration is still correct

### CODE REVIEW

File: enterprise/internal/insights/migration/migration.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+
+	// migrationComplete, workCompleted, err = performUserMigration(tx)
+	// if !migrationComplete || workCompleted {
+	// 	return nil
+	// }
+
+	return nil
+}
+
+// TODO: I don't think we need this at all, do we? What would it do? Should we use it to wipe out the jobs table to restart the migration?
```
Comment (coury-clark) [inline enterprise/internal/insights/migration/migration.go:84]: Right, I think as long as the migration is non-destructive (which this is), we are already backwards compatible and don't need a down direction.

File: enterprise/internal/insights/migration/migration.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+// func performUserMigration(tx *basestore.Store) (bool, bool, error) {
+// 	// This will probably follow the same logic as orgs
+// }
+
+// func performMigrationForRow(tx *basestore.Store, settingsRow someType) (bool, bool, error) {
+// 	// At this point we've picked up the row and are responsible for doing the migration for everything in that settings object.
+// 	// Here I think we can use a good deal of the code that already exists
+
+// 	// Caveats: we can't wipe everything out that already exists. We'll need to match on ids and check if something exists
```
Comment (coury-clark) [inline enterprise/internal/insights/migration/migration.go:221]: Another thing to consider is that currently, the IDs in settings are unique _per cascade_ instead of unique globally. This is probably going to be a massive headache to replicate, so if we can find a simple way around this that would be ideal. I suspect this will mostly be a problem for dashboards referencing ids in parent (shared) levels.

File: migrations/frontend/1528395945_settings_migration_out_of_band.down.sql [hunk 3]
```diff
@@ -0,0 +1,9 @@
+BEGIN;
+
+DROP TABLE IF EXISTS insights_settings_migration_jobs;
+
+DELETE
+FROM out_of_band_migrations
+WHERE id = 14;
```
Comment (coury-clark) [inline migrations/frontend/1528395945_settings_migration_out_of_band.down.sql:7]: This is still good as of right now

File: migrations/frontend/1528395945_settings_migration_out_of_band.up.sql [hunk 4]
```diff
@@ -0,0 +1,51 @@
+BEGIN;
+
+CREATE TABLE IF NOT EXISTS insights_settings_migration_jobs
```
Comment (coury-clark) [inline migrations/frontend/1528395945_settings_migration_out_of_band.up.sql:3]: This migration needs to be number `...947` now

File: migrations/frontend/1528395945_settings_migration_out_of_band.down.sql [hunk 5]
```diff
@@ -0,0 +1,9 @@
+BEGIN;
+
+DROP TABLE IF EXISTS insights_settings_migration_jobs;
```
Comment (coury-clark) [inline migrations/frontend/1528395945_settings_migration_out_of_band.down.sql:3]: This migration needs to be number ...947 now

File: migrations/frontend/1528395945_settings_migration_out_of_band.up.sql [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+FROM settings
+WHERE user_id IS NOT NULL
+ORDER BY user_id, id DESC;
+
+
+INSERT INTO out_of_band_migrations(id, team, component, description, non_destructive,
+                                   apply_reverse, is_enterprise, introduced_version_major, introduced_version_minor)
+VALUES (14, 'code-insights', 'db.insights_settings_migration_jobs',
+        'Migrating insight definitions from settings files to database tables as a last stage to use the GraphQL API.',
+        TRUE, FALSE, TRUE, 3, 34)
```
Comment (coury-clark) [inline migrations/frontend/1528395945_settings_migration_out_of_band.up.sql:48]:
```suggestion
        TRUE, FALSE, TRUE, 3, 35)
```

**Review by coury-clark — CHANGES_REQUESTED** [review]
A few changes required to fix migration numbers, but other than that LGTM

**Review by coury-clark — APPROVED** [review]
LGTM!
Before this is merged, please do a final check on the migration number with the latest tip of `main`.
